include/qphix/clover_product.h

```diff
@@ -128,17 +127,20 @@ struct InnerCloverProduct<
                 cplx_mul_acc(spinor_out[c_out][four_s_out][re][xi],
                              spinor_out[c_out][four_s_out][im][xi],
                              diag_in[sc_in][veclen_idx],
-                             FT{0},
+                             QPhiX::rep<FT, double>(0.0),
                              spinor_in[c_in][four_s_in][re][xi],
                              spinor_in[c_in][four_s_in][im][xi]);
               } else if (sc_out < sc_in) {
                 auto const idx15 = sc_in * (sc_in - 1) / 2 + sc_out;
-                cplx_mul_acc(spinor_out[c_out][four_s_out][re][xi],
-                             spinor_out[c_out][four_s_out][im][xi],
-                             off_diag_in[idx15][re][veclen_idx],
-                             -off_diag_in[idx15][im][veclen_idx],
-                             spinor_in[c_in][four_s_in][re][xi],
-                             spinor_in[c_in][four_s_in][im][xi]);
+                cplx_mul_acc(
+                    spinor_out[c_out][four_s_out][re][xi],
+                    spinor_out[c_out][four_s_out][im][xi],
+                    off_diag_in[idx15][re][veclen_idx],
+                    // aww hell, maybe one should just add negation to QPhiX::half ?
```

Comment:
Since `QPhiX::half` is just `typedef unsigned short half;`, I doubt that we can actually do that. Haskell has the `type` and `newtype` constructs, the latter lets you reinterpret some other type as a new type without actually changing the inner workings. This is nice to have a new `double` type that just compares differently than normal `double`.

In C++, I think we would need to turn `half` into a class and add all those `operator+`, `operator-`, ..., such that it behaves like a numeric type. This would be cleaner because right now we have all those operators, but they don't do what we want.

The Intel intrinsics expect to get a pointer to `uint16_t`, though. So if we have arrays of our `class Half`, we would need to also provide an `operator uint8_t()` such that it could be converted/cast. But this would not allow the conversion `*Half` → `*uint8_t`:

```cpp
#include <cstdint>
#include <iostream>

class Half {
public:
  operator uint16_t() { return data_; }

private:
  uint16_t data_ = -1;
};

void g(uint16_t const h) {
  std::cout << h << std::endl;
}

void f(uint16_t const *const ph) {
  std::cout << (*ph) << std::endl;
}

int main(int argc, char **argv) {
  Half h;
  Half *ph = &h;
  g(h);
  //f(ph);
}
```

When you try to call `f(ph)`, it fails. Therefore we would have to work on the interface to the kernel functions when we do this.

Not sure whether it is worth rewriting in order to remove these subtle bugs (also the whole explicit `rep` stuff would not be needed any more, I think) and potentially other bugs that we have not found yet.

Reply:
No, right now it's not worth it. If Intel ever releases a chip which has native support for half-precision (like GPUs), it will be worth to invest the effort, if we are still using QPhiX at that point.